library.c

```diff
@@ -84,6 +84,27 @@
         } \
     } while (0)
 
+/** Set return value to false in case of we are in atomic mode or add FALSE to output array in pipeline mode */
+#define REDIS_RESPONSE_ERROR(redis_sock, z_tab) \
+    do { \
+        if (IS_ATOMIC(redis_sock)) { \
+            RETVAL_FALSE; \
+        } else { \
+            add_next_index_bool(z_tab, 0); \
+        } \
+    } while (0)
+
+/** Set return value to `zval` in case of we are in atomic mode or add `zval` to output array in pipeline mode */
+#define REDIS_RETURN_ZVAL(redis_sock, z_tab, zval) \
+    do { \
+        if (IS_ATOMIC(redis_sock)) { \
+            /* Move value of `zval` to `return_value` */ \
+            ZVAL_COPY_VALUE(return_value, &zval); \
```

Comment:
Currently there are `RETVAL_ZVAL(&zval, 1, 0)` and `RETVAL_ZVAL(&zval, 0, 0)` used in code
Are you sure we can copy value and it couldn't cause memory leaks?

Reply:
It should be safe, as `add_next_index_zval` method also uses `ZVAL_COPY_VALUE` that moves value from one zval to another one, so there is no reason to call destructor. But I can check it just for sure again. 

https://github.com/php/php-src/blob/7e1d035077d51bd22d3c49ab453f05a2ff23abad/Zend/zend_hash.c#L1185